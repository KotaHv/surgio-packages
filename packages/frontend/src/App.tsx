import React from 'react';
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link as RouterLink,
  LinkProps as RouterLinkProps,
} from 'react-router-dom';
import AppBar from '@material-ui/core/AppBar';
import CssBaseline from '@material-ui/core/CssBaseline';
import Divider from '@material-ui/core/Divider';
import Drawer from '@material-ui/core/Drawer';
import Hidden from '@material-ui/core/Hidden';
import IconButton from '@material-ui/core/IconButton';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import Link from '@material-ui/core/Link';
import ListIcon from '@material-ui/icons/List';
import MenuIcon from '@material-ui/icons/Menu';
import GitHubIcon from '@material-ui/icons/GitHub';
import { makeStyles, useTheme, Theme, createStyles } from '@material-ui/core/styles';

import './App.css';
import ArtifactListPage from './pages/ArtifactList';

const drawerWidth = 240;

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      display: 'flex',
    },
    drawer: {
      [theme.breakpoints.up('md')]: {
        width: drawerWidth,
        flexShrink: 0,
      },
    },
    appBar: {
      [theme.breakpoints.up('md')]: {
        width: `calc(100% - ${drawerWidth}px)`,
        marginLeft: drawerWidth,
      },
    },
    menuButton: {
      marginRight: theme.spacing(2),
      [theme.breakpoints.up('md')]: {
        display: 'none',
      },
    },
    toolbar: {
      ...theme.mixins.toolbar,
      display: 'flex',
    },
    drawerPaper: {
      width: drawerWidth,
    },
    content: {
      width: '100%',
      [theme.breakpoints.up('md')]: {
        width: `calc(100% - ${drawerWidth}px)`,
      },
      padding: theme.spacing(3),
    },
    pageTitle: {
      flexGrow: 1,
    },
    appBarIcons: {
      color: '#fff',
    },
  }),
);

interface ResponsiveDrawerProps {
  container?: Element;
}

export default function App(props: ResponsiveDrawerProps) {
  const { container } = props;
  const classes = useStyles();
  const theme = useTheme();
  const [mobileOpen, setMobileOpen] = React.useState(false);

  const handleDrawerToggle = () => {
    setMobileOpen(!mobileOpen);
  };

  const drawer = (
    <div>
      <div className={classes.toolbar} />
      <Divider />
      <List>
        <ListItemLink to="/list-artifact" primary="Artifacts" icon={<ListIcon />} />
      </List>
    </div>
  );

  return (
    <Router>
      <div className={classes.root}>
        <CssBaseline />
        <AppBar position="fixed" className={classes.appBar}>
          <Toolbar>
            <IconButton
              color="inherit"
              aria-label="open drawer"
              edge="start"
              onClick={handleDrawerToggle}
              className={classes.menuButton}
            >
              <MenuIcon />
            </IconButton>
            <Typography variant="h6" noWrap className={classes.pageTitle}>
              Surgio Dashboard
            </Typography>
            <div>
              <Link href="https://github.com/geekdada/surgio"
                    rel="noopener noreferrer"
                    target="_blank">
                <GitHubIcon className={classes.appBarIcons} />
              </Link>
            </div>
          </Toolbar>
        </AppBar>
        <nav className={classes.drawer} aria-label="folders">
          <Hidden lgUp>
            <Drawer
              container={container}
              variant="temporary"
              open={mobileOpen}
              onClose={handleDrawerToggle}
              classes={{
                paper: classes.drawerPaper,
              }}
              ModalProps={{
                keepMounted: true, // Better open performance on mobile.
              }}
            >
              {drawer}
            </Drawer>
          </Hidden>
          <Hidden smDown>
            <Drawer
              classes={{
                paper: classes.drawerPaper,
              }}
              variant="permanent"
              open
            >
              {drawer}
            </Drawer>
          </Hidden>
        </nav>

        <main className={classes.content}>
          <div className={classes.toolbar} />
          <Switch>
            <Route path="/list-artifact">
              <ArtifactListPage />
            </Route>
          </Switch>
        </main>
      </div>
    </Router>
  );
}

interface ListItemLinkProps {
  icon?: React.ReactElement;
  primary: string;
  to: string;
}

function ListItemLink(props: ListItemLinkProps) {
  const { icon, primary, to } = props;

  const renderLink = React.useMemo(
    () =>
      React.forwardRef<any, Omit<RouterLinkProps, 'to'>>((itemProps, ref) => (
        <RouterLink to={to} ref={ref} {...itemProps} />
      )),
    [to],
  );

  return (
    <li>
      <ListItem button component={renderLink}>
        {icon ? <ListItemIcon>{icon}</ListItemIcon> : null}
        <ListItemText primary={primary} />
      </ListItem>
    </li>
  );
}
