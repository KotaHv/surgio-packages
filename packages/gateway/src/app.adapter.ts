import { FastifyAdapter } from '@nestjs/platform-fastify';
import { ServerFactoryFunction } from 'fastify';

export interface AdapterOptions {
  readonly serverFactory?: ServerFactoryFunction;
}
export function createAdapter(options: AdapterOptions = {}): FastifyAdapter {
  const adapter = new FastifyAdapter({
    serverFactory: options.serverFactory,
    logger: false,
    trustProxy: true,
  });

  return adapter;
}
