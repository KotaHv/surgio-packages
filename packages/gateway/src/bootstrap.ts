import { NestFactory } from '@nestjs/core';
import {
  FastifyAdapter,
  NestFastifyApplication,
} from '@nestjs/platform-fastify';
import FastifyCookie from 'fastify-cookie';

import { AppModule } from './app.module';
import { HttpExceptionFilter } from './filter/http-exception.filter';

export async function bootstrap(): Promise<void> {
  const app = await NestFactory.create<NestFastifyApplication>(
    AppModule,
    new FastifyAdapter({
      logger: {
        level: 'warn',
        serializers: {
          req(req: any): any {
            return {
              method: req.method,
              url: req.url,
              'user-agent': req.headers['user-agent'] || '-',
              hostname: req.hostname,
              remoteAddress: req.ip,
              remotePort: req.connection.remotePort
            }
          }
        },
        prettyPrint: true,
      },
    })
  );
  const configService = app.get('ConfigService');
  const port = configService.get('port');
  const secret = configService.get('secret');

  app.useGlobalFilters(new HttpExceptionFilter());

  app.register(FastifyCookie, {
    secret, // for cookies signature
    parseOptions: {}     // options for parsing cookies
  });

  await app.listen(port);

  console.log('> Your app is ready at http://127.0.0.1:' + port);
}
