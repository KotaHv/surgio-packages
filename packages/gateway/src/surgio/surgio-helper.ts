import { basename, join } from 'path';
import fs, { promises as fsp } from 'fs';
import { Environment } from 'nunjucks';
import semver from 'semver';
import { Logger } from '@nestjs/common';
import { getEngine } from 'surgio/build/generator/template';
import { getProvider } from 'surgio/build/provider';
import { ArtifactConfig, CommandConfig, RemoteSnippet } from 'surgio/build/types';
import { loadRemoteSnippetList } from 'surgio/build/utils/remote-snippet';
import { PackageJson } from 'type-fest';

type PossibleProviderType = ReturnType<typeof getProvider>;

export class SurgioHelper {
  public remoteSnippetList: ReadonlyArray<RemoteSnippet>;
  public artifactList: ReadonlyArray<ArtifactConfig>;
  public providerMap: Map<string, PossibleProviderType> = new Map();
  public readonly templateEngine?: Environment;

  private readonly pkgFile?: PackageJson;

  constructor(public cwd: string, public readonly config: CommandConfig) {
    const pkgFile = join(cwd, 'package.json');

    this.artifactList = config.artifacts;
    this.templateEngine = getEngine(config.templateDir, config.publicUrl);
    if (fs.existsSync(pkgFile)) {
      this.pkgFile = require(pkgFile);
    }
  }

  public async init(): Promise<this> {
    await this.checkCoreVersion();

    const remoteSnippetsConfig = this.config.remoteSnippets || [];
    this.remoteSnippetList = await loadRemoteSnippetList(remoteSnippetsConfig);

    await this.readProviders();

    return this;
  }

  private async readProviders(): Promise<void> {
    const files = await fsp.readdir(this.config.providerDir, {
      encoding: 'utf8',
    });

    async function readProvider(path): Promise<PossibleProviderType> {
      let provider;

      try {
        const providerName = basename(path, '.js');

        provider = getProvider(providerName, require(path));
      } catch (err) {
        return undefined;
      }

      if (!provider?.type) {
        return undefined;
      }

      return provider;
    }

    for (const file of files) {
      const result = await readProvider(join(this.config.providerDir, file));
      if (result) {
        this.providerMap.set(result.name, result);
      }
    }
  }

  private async checkCoreVersion(): Promise<void> {
    const corePkgFile = require('surgio/package.json');
    const gatewayPkgFile = require('../../package.json');
    const peerVersion = gatewayPkgFile.peerDependencies.surgio;

    if (!semver.satisfies(corePkgFile.version, peerVersion)) {
      Logger.warn('', undefined, false);
      Logger.warn('Surgio 版本过低，请升级后重新运行！', undefined, false);
      Logger.warn('', undefined, false);
      Logger.warn('  命令：', undefined, false);
      Logger.warn('  npm install surgio@latest', undefined, false);
      Logger.warn('', undefined, false);
      throw new Error('Surgio 版本过低');
    }
  }
}
