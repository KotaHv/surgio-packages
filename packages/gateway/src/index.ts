import 'source-map-support/register';
import { ConfigService } from '@nestjs/config';
import { NowRequest, NowResponse } from '@vercel/node/dist';
import { createServer, Server } from 'http';

import { bootstrap } from './bootstrap';

export const createHttpServer = (): Server => {
  const prepare = bootstrap()
    .then(nestApp => {
      return nestApp.init();
    });

  return createServer((req: NowRequest, res: NowResponse) => {
    prepare.then(nestApp => {
      nestApp.getHttpServer().emit('request', req, res);
    });
  });
};

export const startServer = (): Promise<Server> => {
  let port: number;

  return bootstrap()
    .then(app => {
      const configService = app.get<ConfigService>('ConfigService');
      port = configService.get('port') as number;

      return app.listen(port);
    })
    .then((httpServer: Server) => {
      console.log('> Your app is ready at http://127.0.0.1:' + port);
      return httpServer;
    })
    .catch(err => {
      console.error('⚠️ 服务启动失败');
      console.error(err);
      process.exit(1);
    });
};

// export const createAwsHandler = (): ((event: any, context: Context) => void) => {
//   let lambdaProxy: Server;
//   const prepare = bootstrap()
//     .then(nestApp => {
//       return nestApp.init();
//     });
//
//   return (event: any, context: Context) => {
//     if (!lambdaProxy) {
//       prepare.then((nestApp) => {
//         lambdaProxy = nestApp.getHttpServer();
//         serverlessExpress.proxy(lambdaProxy, event, context);
//       });
//     } else {
//       serverlessExpress.proxy(lambdaProxy, event, context);
//     }
//   };
// };
//
// export const createServerlessHanlder = () => {
//   let lambdaProxy: ServerlessHandler;
//   const prepare = bootstrap()
//     .then(nestApp => {
//       return nestApp.init();
//     });
//
//   return async (event: any, context: Context): Promise<APIGatewayProxyResult> => {
//     if (!lambdaProxy) {
//       const nestApp = await prepare;
//       lambdaProxy = serverless(nestApp.getHttpServer());
//       return await lambdaProxy(event, context);
//     } else {
//       return await lambdaProxy(event, context);
//     }
//   };
// };

export const bootstrapServer = bootstrap;
