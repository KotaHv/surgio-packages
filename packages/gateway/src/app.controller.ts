import { Controller, Get, Res, Param, Query, HttpException, HttpStatus, UseGuards, Req, Logger } from '@nestjs/common';
import { FastifyReply, FastifyRequest } from 'fastify';
import { ServerResponse } from 'http';
import { Artifact } from 'surgio/build/generator/artifact';
import _ from 'lodash';
import { URL } from 'url';

import { BearerAuthGuard } from './auth/bearer.guard';
import { SurgioService } from './surgio/surgio.service';

@Controller()
export class AppController {
  private readonly logger = new Logger(AppController.name);

  constructor(private readonly surgioService: SurgioService) {}

  @UseGuards(BearerAuthGuard)
  @Get('/get-artifact/:name')
  public async getArtifact(
    @Res() res: FastifyReply<ServerResponse>,
    @Param() params,
    @Query() query: GetArtifactQuery,
    @Req() req: FastifyRequest
  ): Promise<void> {
    const dl = query.dl;
    const format = query.format;
    const filter = query.filter;
    const urlParams = _.omit(query, ['dl', 'format', 'filter', 'access_token']);
    const artifactName: string = params.name;
    const artifact = format !== void 0 ?
      await this.surgioService.transformArtifact(artifactName, format, filter) :
      await this.surgioService.getArtifact(artifactName);

    if (artifact) {
      res.header('content-type', 'text/plain; charset=utf-8');
      res.header('cache-control', 'private, no-cache, no-stores');

      if (dl === '1') {
        res.header('content-disposition', `attachment; filename="${artifactName}"`);
      }

      this.logger.warn(`[download-artifact] ${artifactName} ${req.headers['user-agent'] || '-'}`);

      // @ts-ignore
      await this.sendPayload(req, res, artifact, urlParams);
    } else {
      throw new HttpException('NOT FOUND', HttpStatus.NOT_FOUND);
    }
  }

  @UseGuards(BearerAuthGuard)
  @Get('/export-providers')
  public async exportProvider(
    @Req() req: FastifyRequest,
    @Res() res: FastifyReply<ServerResponse>,
    @Query() query: ExportProviderQuery,
  ): Promise<void> {
    const providers: string[] = query.providers ? query.providers.split(',').map(item => item.trim()) : [];

    if (!providers.length) {
      throw new HttpException('参数 provider 必须指定至少一个值', HttpStatus.BAD_REQUEST);
    }

    providers.forEach(provider => {
      if (!this.surgioService.surgioHelper.providerMap.has(provider)) {
        throw new HttpException(`provider ${provider} 不存在`, HttpStatus.NOT_FOUND);
      }
    });

    const format = query.format;
    const template = query.template;

    if (!format && !template) {
      throw new HttpException('参数 format 和 template 必须指定至少一个值', HttpStatus.BAD_REQUEST);
    }

    const dl = query.dl;
    const filter = query.filter;
    const urlParams = _.omit(query, ['dl', 'format', 'filter', 'access_token', 'providers']);
    let artifact: Artifact;

    if (format) {
      artifact = await this.surgioService.exportProvider(providers[0], format, undefined, {
        filter,
        ...(providers.length > 1 ? {
          combineProviders: providers.splice(1),
        } : null),
      });
    } else {
      artifact = await this.surgioService.exportProvider(providers[0], undefined, template, {
        filter,
        downloadUrl: (new URL(req.req.url as string, this.surgioService.config.publicUrl)).toString(),
        ...(providers.length > 1 ? {
          combineProviders: providers.splice(1),
        } : null),
      });
    }

    if (artifact) {
      res.header('content-type', 'text/plain; charset=utf-8');
      res.header('cache-control', 'private, no-cache, no-stores');

      if (dl === '1') {
        res.header('content-disposition', `attachment; filename="${artifact.artifact.name}"`);
      }

      this.logger.warn(`[download-artifact] ${artifact.artifact.name} ${req.headers['user-agent'] || '-'}`);

      // @ts-ignore
      await this.sendPayload(req, res, artifact, urlParams);
    } else {
      throw new HttpException('NOT FOUND', HttpStatus.NOT_FOUND);
    }
  }

  private async sendPayload(
    req: FastifyRequest,
    res: FastifyReply<ServerResponse>,
    artifact: string|Artifact,
    urlParams?: Record<string, string>,
  ): Promise<void> {
    if (typeof artifact === 'string') {
      res.send(artifact);
    } else {
      // 只支持输出单个 Provider 的流量信息
      if (artifact.providerMap.size === 1) {
        const providers = artifact.providerMap.values();
        const provider = providers.next().value;

        if (provider.supportGetSubscriptionUserInfo) {
          const subscriptionUserInfo = await provider.getSubscriptionUserInfo();

          if (subscriptionUserInfo) {
            const values = ['upload', 'download', 'total', 'expire']
              .map(key => `${key}=${subscriptionUserInfo[key] || 0}`);

            res.header(
              'subscription-userinfo',
              values.join('; ')
            );
          }
        }
      }

      res.send(artifact.render(
        undefined,
        {
          urlParams: urlParams ? this.processUrlParams(urlParams) : undefined,
        })
      );
    }
  }

  private processUrlParams(urlParams: Record<string, string>): object {
    const result: NonNullable<any> = Object.create(null);

    Object.keys(urlParams)
      .forEach(key => {
        _.set(result, key, urlParams[key]);
      });

    return result;
  }
}

interface GetArtifactQuery {
  readonly format: string;
  readonly filter?: string;
  readonly dl?: string;
  readonly access_token?: string;
  readonly [key: string]: string|undefined;
}

interface ExportProviderQuery {
  readonly providers: string;
  readonly format?: string;
  readonly template?: string;
  readonly filter?: string;
  readonly dl?: string;
  readonly access_token?: string;
  readonly [key: string]: string|undefined;
}
