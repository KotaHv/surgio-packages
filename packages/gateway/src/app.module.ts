import { Module } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { join } from 'path';
import { ServeStaticModule } from '@nestjs/serve-static';

import { AppController } from './app.controller';
import { ApiModule } from './api/api.module';
import { SurgioModule } from './surgio/surgio.module';
import { SurgioService } from './surgio/surgio.service';
import { AuthModule } from './auth/auth.module';
import configuration from './config/configuration';

const CWD = process.env.SURGIO_PROJECT_DIR || process.cwd();
const FE_FOLDER = require.resolve('@surgio/gateway-frontend');

@Module({
  imports: [
    ServeStaticModule.forRoot({
      rootPath: join(FE_FOLDER, 'build'),
    }),
    ConfigModule.forRoot({
      ignoreEnvFile: true,
      isGlobal: true,
      load: [configuration],
    }),
    SurgioModule.register({
      cwd: CWD,
    }),
    ApiModule,
    AuthModule,
  ],
  controllers: [AppController],
  providers: [SurgioService],
})
export class AppModule {}
