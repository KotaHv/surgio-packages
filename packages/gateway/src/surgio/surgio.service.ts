import { Injectable, Inject, HttpException, HttpStatus } from '@nestjs/common';
import { generate } from 'surgio/build/generate';
import { Artifact } from 'surgio/build/generator/artifact';
import { getProvider } from 'surgio/build/provider';
import { ArtifactConfig } from 'surgio/build/types';
import * as filters from 'surgio/build/utils/filter';

import { SurgioHelper } from './surgio-helper';

@Injectable()
export class SurgioService {
  constructor(@Inject('SURGIO_HELPER') public surgioHelper: SurgioHelper) {}

  public async getArtifact(artifactName: string): Promise<Artifact> {
    const target = this.surgioHelper.artifactList.filter(item => item.name === artifactName);

    if (!target.length) {
      return undefined;
    }

    const artifactInstance = new Artifact(
      this.surgioHelper.config,
      target[0],
      {
        remoteSnippetList: this.surgioHelper.remoteSnippetList,
      }
    );

    return await artifactInstance.init(this.surgioHelper.templateEngine);
  }

  public async transformArtifact(artifactName: string, format: string, filter?: string): Promise<Artifact|string|HttpException> {
    const target = this.surgioHelper.artifactList.filter(item => item.name === artifactName);
    let filterName;

    if (!target.length) {
      return undefined;
    }
    if (filter) {
      filterName = filters.hasOwnProperty(filter) ? filter : `customFilters.${filter}`;
    }

    switch (format) {
      case 'surge-policy': {
        const artifact = {
          ...target[0],
          template: undefined,
          templateString: `{{ getSurgeNodes(nodeList${filterName ? `, ${filterName}` : ''}) }}`,
        };
        return await generate(
          this.surgioHelper.config,
          artifact,
          this.surgioHelper.remoteSnippetList,
          this.surgioHelper.templateEngine
        );
      }

      case 'qx-server': {
        const artifact = {
          ...target[0],
          template: undefined,
          templateString: `{{ getQuantumultXNodes(nodeList${filterName ? `, ${filterName}` : ''}) }}`,
        };
        return await generate(
          this.surgioHelper.config,
          artifact,
          this.surgioHelper.remoteSnippetList,
          this.surgioHelper.templateEngine
        );
      }

      case 'clash-provider': {
        const artifact = {
          ...target[0],
          template: undefined,
          templateString: [
            '---',
            'proxies:',
            `{{ getClashNodes(nodeList${filterName ? `, ${filterName}` : ''}) | yaml }}`,
            '...'
          ].join('\n')
        };
        return await generate(
          this.surgioHelper.config,
          artifact,
          this.surgioHelper.remoteSnippetList,
          this.surgioHelper.templateEngine
        );
      }

      default:
        return new HttpException( 'unsupported format', HttpStatus.BAD_REQUEST);
    }
  }

  public listProviders(): ReadonlyArray<ReturnType<typeof getProvider>> {
    return Array.from(this.surgioHelper.providerMap.values());
  }
}
