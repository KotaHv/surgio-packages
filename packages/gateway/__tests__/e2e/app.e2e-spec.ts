import { NestFastifyApplication } from '@nestjs/platform-fastify';

import { bootstrap } from '../../src/bootstrap';
import { SurgioService } from '../../src/surgio/surgio.service';

describe('AppController (e2e)', () => {
  let app: NestFastifyApplication;
  let token;

  beforeAll(async () => {
    app = await bootstrap();

    const surgioService = app.get<SurgioService>('SurgioService');
    token = surgioService.config.gateway?.accessToken;

    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  test('/ (GET)', async () => {
    const res = await app.inject({
      url: '/',
    });

    expect(res.statusCode).toBe(200);
  });

  test('/get-artifact (GET)', async () => {
    const res = await app.inject({
      url: '/get-artifact/test.conf',
      query: {
        access_token: token,
      },
    });

    expect(res.statusCode).toBe(200);
    expect(res.headers['subscription-userinfo']).toBeUndefined();
    expect(res.payload).toMatchSnapshot();
  });

  test('/get-artifact (GET) attachment', async () => {
    const res = await app.inject({
      url: '/get-artifact/test.conf',
      query: {
        access_token: token,
        dl: '1',
      },
    });

    expect(res.headers['content-disposition']).toBe('attachment; filename="test.conf"');
  });

  test('/get-artifact (GET) userinfo header', async () => {
    const res = await app.inject({
      url: '/get-artifact/test3.conf',
      query: {
        access_token: token,
      },
    });

    expect(res.headers['subscription-userinfo'])
      .toBe('upload=891332010; download=29921186546; total=322122547200; expire=1586330887');
  });

  test('/get-artifact (GET) 404', async () => {
    const res = await app.inject({
      url: '/get-artifact/notfound.conf',
      query: {
        access_token: token,
      },
    });

    expect(res.statusCode).toBe(404);
  });

  test('/get-artifact (GET) unauthorized request', async () => {
    expect((
      await app.inject({
        url: '/get-artifact/test.conf',
      })
    ).statusCode).toBe(401);
    expect((
      await app.inject({
        url: '/get-artifact/test.conf',
        query: {
          access_token: 'wrong',
        },
      })
    ).statusCode).toBe(401);
  });
});
