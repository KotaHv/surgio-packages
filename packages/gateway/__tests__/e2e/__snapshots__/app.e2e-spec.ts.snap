// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController (e2e) /export-providers (GET) 1`] = `
"ss1 = custom, server, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, udp-relay=true
ss2 = custom, server, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, udp-relay=false, obfs=tls, obfs-host=www.bing.com
vmess = vmess, server, 443, username=uuid
http 1 = https, server, 443, username, password, skip-cert-verify=false
http 2 = http, server, 443, username, password
snell = snell, server, 44046, psk=yourpsk, obfs=http
ss4 = custom, server, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, udp-relay=false, obfs=tls, obfs-host=example.com"
`;

exports[`AppController (e2e) /export-providers (GET) global filter 1`] = `"🇺🇲 US = custom, us.example.com, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, udp-relay=true, obfs=tls, obfs-host=gateway-carry.icloud.com, tfo=true"`;

exports[`AppController (e2e) /export-providers (GET) internal filter 1`] = `""`;

exports[`AppController (e2e) /export-providers (GET) multiple providers 1`] = `
"ss1 = custom, server, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, udp-relay=true
ss2 = custom, server, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, udp-relay=false, obfs=tls, obfs-host=www.bing.com
vmess = vmess, server, 443, username=uuid
http 1 = https, server, 443, username, password, skip-cert-verify=false
http 2 = http, server, 443, username, password
snell = snell, server, 44046, psk=yourpsk, obfs=http
ss4 = custom, server, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, udp-relay=false, obfs=tls, obfs-host=example.com
🇺🇲 US = custom, us.example.com, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, udp-relay=true, obfs=tls, obfs-host=gateway-carry.icloud.com, tfo=true
Snell = snell, us.example.com, 443, psk=password, obfs=tls
HTTPS = https, us.example.com, 443, username, password"
`;

exports[`AppController (e2e) /export-providers (GET) multiple providers 404 1`] = `"{\\"statusCode\\":404,\\"error\\":\\"provider notfound 不存在\\"}"`;

exports[`AppController (e2e) /export-providers (GET) private filter 1`] = `
"ss1 = custom, server, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, udp-relay=true
ss2 = custom, server, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, udp-relay=false, obfs=tls, obfs-host=www.bing.com
vmess = vmess, server, 443, username=uuid
ss4 = custom, server, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, udp-relay=false, obfs=tls, obfs-host=example.com
🇺🇲 US = custom, us.example.com, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, udp-relay=true, obfs=tls, obfs-host=gateway-carry.icloud.com, tfo=true"
`;

exports[`AppController (e2e) /get-artifact (GET) 1`] = `
"🇺🇸US 1 = custom, us.example.com, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, obfs=tls, obfs-host=gateway-carry.icloud.com
🇺🇸US 2 = custom, us.example.com, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module
🇺🇲 US = custom, us.example.com, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, udp-relay=true, obfs=tls, obfs-host=gateway-carry.icloud.com, tfo=true
Snell = snell, us.example.com, 443, psk=password, obfs=tls
HTTPS = https, us.example.com, 443, username, password
ss1 = custom, server, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, udp-relay=true
ss2 = custom, server, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, udp-relay=false, obfs=tls, obfs-host=www.bing.com
vmess = vmess, server, 443, username=uuid
http 1 = https, server, 443, username, password, skip-cert-verify=false
http 2 = http, server, 443, username, password
snell = snell, server, 44046, psk=yourpsk, obfs=http
ss4 = custom, server, 443, chacha20-ietf-poly1305, password, https://raw.githubusercontent.com/ConnersHua/SSEncrypt/master/SSEncrypt.module, udp-relay=false, obfs=tls, obfs-host=example.com
----
🇺🇸US 1, 🇺🇸US 2, 🇺🇸US 3, 🇺🇲 US, Snell, HTTPS, ss1, ss2, ss3, vmess, http 1, http 2, snell, ss4, ss-wss
"
`;
